<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad.sample_qc.relatedness &mdash; gnomad master documentation</title><link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="gnomad.sample_qc.sex" href="sex.html" />
    <link rel="prev" title="gnomad.sample_qc.platform" href="platform.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> gnomad
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples/index.html">Examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../variant_qc/index.html">variant_qc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../assessment/index.html">assessment</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">sample_qc</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="ancestry.html">ancestry</a></li>
<li class="toctree-l3"><a class="reference internal" href="filtering.html">filtering</a></li>
<li class="toctree-l3"><a class="reference internal" href="pipeline.html">pipeline</a></li>
<li class="toctree-l3"><a class="reference internal" href="platform.html">platform</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">relatedness</a></li>
<li class="toctree-l3"><a class="reference internal" href="sex.html">sex</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../resources/index.html">resources</a></li>
<li class="toctree-l2"><a class="reference internal" href="../utils/index.html">utils</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../resource_sources.html">Resource Sources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../changelog.html">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">gnomad</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../index.html">gnomad</a> &raquo;</li>
          <li><a href="index.html">gnomad.sample_qc</a> &raquo;</li>
      <li>gnomad.sample_qc.relatedness</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/api_reference/sample_qc/relatedness.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="gnomad-sample-qc-relatedness">
<h1>gnomad.sample_qc.relatedness<a class="headerlink" href="#gnomad-sample-qc-relatedness" title="Permalink to this headline">¶</a></h1>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.UNRELATED" title="gnomad.sample_qc.relatedness.UNRELATED"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.UNRELATED</span></code></a></p></td>
<td><p>String representation for a pair of unrelated individuals in this module.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.SECOND_DEGREE_RELATIVES" title="gnomad.sample_qc.relatedness.SECOND_DEGREE_RELATIVES"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.SECOND_DEGREE_RELATIVES</span></code></a></p></td>
<td><p>String representation for a pair of 2nd degree relatives in this module.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.PARENT_CHILD" title="gnomad.sample_qc.relatedness.PARENT_CHILD"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.PARENT_CHILD</span></code></a></p></td>
<td><p>String representation for a parent-child pair in this module.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.SIBLINGS" title="gnomad.sample_qc.relatedness.SIBLINGS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.SIBLINGS</span></code></a></p></td>
<td><p>String representation for a sibling pair in this module.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.DUPLICATE_OR_TWINS" title="gnomad.sample_qc.relatedness.DUPLICATE_OR_TWINS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.DUPLICATE_OR_TWINS</span></code></a></p></td>
<td><p>String representation for a pair of samples who are identical (either MZ twins of duplicate) in this module.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.AMBIGUOUS_RELATIONSHIP" title="gnomad.sample_qc.relatedness.AMBIGUOUS_RELATIONSHIP"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.AMBIGUOUS_RELATIONSHIP</span></code></a></p></td>
<td><p>String representation for a pair of samples whose relationship is ambiguous.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.get_duplicated_samples" title="gnomad.sample_qc.relatedness.get_duplicated_samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.get_duplicated_samples</span></code></a>(…)</p></td>
<td><p>Extract the list of duplicate samples using a Table ouput from pc_relate.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.get_duplicated_samples_ht" title="gnomad.sample_qc.relatedness.get_duplicated_samples_ht"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.get_duplicated_samples_ht</span></code></a>(…)</p></td>
<td><p>Create a HT with duplicated samples sets.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.explode_duplicate_samples_ht" title="gnomad.sample_qc.relatedness.explode_duplicate_samples_ht"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.explode_duplicate_samples_ht</span></code></a>(dups_ht)</p></td>
<td><p>Explode the result of <cite>get_duplicated_samples_ht</cite>, so that each line contains a single sample.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.get_relationship_expr" title="gnomad.sample_qc.relatedness.get_relationship_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.get_relationship_expr</span></code></a>(…)</p></td>
<td><p>Return an expression indicating the relationship between a pair of samples given their kin coefficient and IBDO, IBD1, IBD2 values.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.infer_families" title="gnomad.sample_qc.relatedness.infer_families"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.infer_families</span></code></a>(…)</p></td>
<td><p>Generate a pedigree containing trios inferred from the <cite>relationship_ht</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.create_fake_pedigree" title="gnomad.sample_qc.relatedness.create_fake_pedigree"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.create_fake_pedigree</span></code></a>(n, …)</p></td>
<td><p>Generate a pedigree made of trios created by sampling 3 random samples in the sample list.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.compute_related_samples_to_drop" title="gnomad.sample_qc.relatedness.compute_related_samples_to_drop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.compute_related_samples_to_drop</span></code></a>(…)</p></td>
<td><p>Compute a Table with the list of samples to drop (and their global rank) to get the maximal independent set of unrelated samples.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.filter_mt_to_trios" title="gnomad.sample_qc.relatedness.filter_mt_to_trios"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.filter_mt_to_trios</span></code></a>(mt, …)</p></td>
<td><p>Filter a MatrixTable to a set of trios in <cite>fam_ht</cite> and annotates with adj.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.generate_trio_stats_expr" title="gnomad.sample_qc.relatedness.generate_trio_stats_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.generate_trio_stats_expr</span></code></a>(trio_mt)</p></td>
<td><p>Generate a row-wise expression containing trio transmission stats.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.sample_qc.relatedness.generate_sib_stats_expr" title="gnomad.sample_qc.relatedness.generate_sib_stats_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.sample_qc.relatedness.generate_sib_stats_expr</span></code></a>(mt, …)</p></td>
<td><p>Generate a row-wise expression containing the number of alternate alleles in common between sibling pairs.</p></td>
</tr>
</tbody>
</table>
<span class="target" id="module-gnomad.sample_qc.relatedness"></span><dl class="data">
<dt id="gnomad.sample_qc.relatedness.UNRELATED">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">UNRELATED</code><em class="property"> = 'unrelated'</em><a class="headerlink" href="#gnomad.sample_qc.relatedness.UNRELATED" title="Permalink to this definition">¶</a></dt>
<dd><p>String representation for a pair of unrelated individuals in this module.
Typically &gt;2nd degree relatives, but the threshold is user-dependant.</p>
</dd></dl>

<dl class="data">
<dt id="gnomad.sample_qc.relatedness.SECOND_DEGREE_RELATIVES">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">SECOND_DEGREE_RELATIVES</code><em class="property"> = 'second degree relatives'</em><a class="headerlink" href="#gnomad.sample_qc.relatedness.SECOND_DEGREE_RELATIVES" title="Permalink to this definition">¶</a></dt>
<dd><p>String representation for a pair of 2nd degree relatives in this module.</p>
</dd></dl>

<dl class="data">
<dt id="gnomad.sample_qc.relatedness.PARENT_CHILD">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">PARENT_CHILD</code><em class="property"> = 'parent-child'</em><a class="headerlink" href="#gnomad.sample_qc.relatedness.PARENT_CHILD" title="Permalink to this definition">¶</a></dt>
<dd><p>String representation for a parent-child pair in this module.</p>
</dd></dl>

<dl class="data">
<dt id="gnomad.sample_qc.relatedness.SIBLINGS">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">SIBLINGS</code><em class="property"> = 'siblings'</em><a class="headerlink" href="#gnomad.sample_qc.relatedness.SIBLINGS" title="Permalink to this definition">¶</a></dt>
<dd><p>String representation for a sibling pair in this module.</p>
</dd></dl>

<dl class="data">
<dt id="gnomad.sample_qc.relatedness.DUPLICATE_OR_TWINS">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">DUPLICATE_OR_TWINS</code><em class="property"> = 'duplicate/twins'</em><a class="headerlink" href="#gnomad.sample_qc.relatedness.DUPLICATE_OR_TWINS" title="Permalink to this definition">¶</a></dt>
<dd><p>String representation for a pair of samples who are identical (either MZ twins of duplicate) in this module.</p>
</dd></dl>

<dl class="data">
<dt id="gnomad.sample_qc.relatedness.AMBIGUOUS_RELATIONSHIP">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">AMBIGUOUS_RELATIONSHIP</code><em class="property"> = 'ambiguous'</em><a class="headerlink" href="#gnomad.sample_qc.relatedness.AMBIGUOUS_RELATIONSHIP" title="Permalink to this definition">¶</a></dt>
<dd><p>String representation for a pair of samples whose relationship is ambiguous.
This is used in the case of a pair of samples which kinship/IBD values do not correspond to any biological relationship between two individuals.</p>
</dd></dl>

<dl class="function">
<dt id="gnomad.sample_qc.relatedness.get_duplicated_samples">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">get_duplicated_samples</code><span class="sig-paren">(</span><em class="sig-param">relationship_ht</em>, <em class="sig-param">i_col='i'</em>, <em class="sig-param">j_col='j'</em>, <em class="sig-param">rel_col='relationship'</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/sample_qc/relatedness.html#get_duplicated_samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.sample_qc.relatedness.get_duplicated_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract the list of duplicate samples using a Table ouput from pc_relate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>relationship_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table with relationships between pairs of samples</p></li>
<li><p><strong>i_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Column containing the 1st sample</p></li>
<li><p><strong>j_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Column containing the 2nd sample</p></li>
<li><p><strong>rel_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Column containing the sample pair relationship annotated with get_relationship_expr</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Set</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>List of sets of samples that are duplicates</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.sample_qc.relatedness.get_duplicated_samples_ht">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">get_duplicated_samples_ht</code><span class="sig-paren">(</span><em class="sig-param">duplicated_samples</em>, <em class="sig-param">samples_rankings_ht</em>, <em class="sig-param">rank_ann='rank'</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/sample_qc/relatedness.html#get_duplicated_samples_ht"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.sample_qc.relatedness.get_duplicated_samples_ht" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a HT with duplicated samples sets.</p>
<p>Each row is indexed by the sample that is kept and also contains the set of duplicate samples that should be filtered.</p>
<p><cite>samples_rankings_ht</cite> is a HT containing a global rank for each of the samples (smaller is better).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>duplicated_samples</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Set</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – List of sets of duplicated samples</p></li>
<li><p><strong>samples_rankings_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – HT with global rank for each sample</p></li>
<li><p><strong>rank_ann</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Annotation in <cite>samples_ranking_ht</cite> containing each sample global rank (smaller is better).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>HT with duplicate sample sets, including which to keep/filter</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.sample_qc.relatedness.explode_duplicate_samples_ht">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">explode_duplicate_samples_ht</code><span class="sig-paren">(</span><em class="sig-param">dups_ht</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/sample_qc/relatedness.html#explode_duplicate_samples_ht"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.sample_qc.relatedness.explode_duplicate_samples_ht" title="Permalink to this definition">¶</a></dt>
<dd><p>Explode the result of <cite>get_duplicated_samples_ht</cite>, so that each line contains a single sample.</p>
<p>An additional annotation is added: <cite>dup_filtered</cite> indicating which of the duplicated samples was kept.
Requires a field <cite>filtered</cite> which type should be the same as the input duplicated samples Table key.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>dups_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Input HT</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Flattened HT</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.sample_qc.relatedness.get_relationship_expr">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">get_relationship_expr</code><span class="sig-paren">(</span><em class="sig-param">kin_expr</em>, <em class="sig-param">ibd0_expr</em>, <em class="sig-param">ibd1_expr</em>, <em class="sig-param">ibd2_expr</em>, <em class="sig-param">first_degree_kin_thresholds=(0.19</em>, <em class="sig-param">0.4)</em>, <em class="sig-param">second_degree_min_kin=0.1</em>, <em class="sig-param">ibd0_0_max=0.025</em>, <em class="sig-param">ibd0_25_thresholds=(0.1</em>, <em class="sig-param">0.425)</em>, <em class="sig-param">ibd1_0_thresholds=(-0.15</em>, <em class="sig-param">0.1)</em>, <em class="sig-param">ibd1_50_thresholds=(0.275</em>, <em class="sig-param">0.75)</em>, <em class="sig-param">ibd1_100_min=0.75</em>, <em class="sig-param">ibd2_0_max=0.125</em>, <em class="sig-param">ibd2_25_thresholds=(0.1</em>, <em class="sig-param">0.5)</em>, <em class="sig-param">ibd2_100_thresholds=(0.75</em>, <em class="sig-param">1.25)</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/sample_qc/relatedness.html#get_relationship_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.sample_qc.relatedness.get_relationship_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an expression indicating the relationship between a pair of samples given their kin coefficient and IBDO, IBD1, IBD2 values.</p>
<p>The kinship coefficient values in the defaults are in line with those output from
<cite>hail.methods.pc_relate &lt;https://hail.is/docs/0.2/methods/genetics.html?highlight=pc_relate#hail.methods.pc_relate&gt;</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kin_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>) – Kin coefficient expression</p></li>
<li><p><strong>ibd0_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>) – IBDO expression</p></li>
<li><p><strong>ibd1_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>) – IBD1 expression</p></li>
<li><p><strong>ibd2_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>) – IDB2 expression</p></li>
<li><p><strong>first_degree_kin_thresholds</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – (min, max) kinship threshold for 1st degree relatives</p></li>
<li><p><strong>second_degree_min_kin</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – min kinship threshold for 2nd degree relatives</p></li>
<li><p><strong>ibd0_0_max</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – max IBD0 threshold for 0 IBD0 sharing</p></li>
<li><p><strong>ibd0_25_thresholds</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – (min, max) thresholds for 0.25 IBD0 sharing</p></li>
<li><p><strong>ibd1_0_thresholds</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – (min, max) thresholds for 0 IBD1 sharing. Note that the min is there because pc_relate can output large negative values in some corner cases.</p></li>
<li><p><strong>ibd1_50_thresholds</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – (min, max) thresholds for 0.5 IBD1 sharing</p></li>
<li><p><strong>ibd1_100_min</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – min IBD1 threshold for 1.0 IBD1 sharing</p></li>
<li><p><strong>ibd2_0_max</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – max IBD2 threshold for 0 IBD2 sharing</p></li>
<li><p><strong>ibd2_25_thresholds</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – (min, max) thresholds for 0.25 IBD2 sharing</p></li>
<li><p><strong>ibd2_100_thresholds</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – (min, max) thresholds for 1.00 IBD2 sharing. Note that the min is there because pc_relate can output much larger IBD2 values in some corner cases.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The relationship annotation using the constants defined in this module.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.sample_qc.relatedness.infer_families">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">infer_families</code><span class="sig-paren">(</span><em class="sig-param">relationship_ht</em>, <em class="sig-param">sex</em>, <em class="sig-param">duplicate_samples_ht</em>, <em class="sig-param">i_col='i'</em>, <em class="sig-param">j_col='j'</em>, <em class="sig-param">relationship_col='relationship'</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/sample_qc/relatedness.html#infer_families"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.sample_qc.relatedness.infer_families" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a pedigree containing trios inferred from the <cite>relationship_ht</cite>.</p>
<p>This function takes a hail Table with a row for each pair of related individuals i, j in the data (it’s OK to have
unrelated samples too).</p>
<dl class="simple">
<dt>The <cite>relationship_col</cite> should be a column specifying the relationship between each two samples as defined in this</dt><dd><p>module’s constants.</p>
</dd>
</dl>
<p>This function returns a pedigree containing trios inferred from the data. Family ID can be the same for multiple
trios if one or more members of the trios are related (e.g. sibs, multi-generational family). Trios are ordered by family ID.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function only returns complete trios defined as: one child, one father and one mother (sex is required for both parents).</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>relationship_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Input relationship table</p></li>
<li><p><strong>sex</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>]]) – A Table or dict giving the sex for each sample (<cite>TRUE`=female, `FALSE`=male). If a Table is given, it should have a field `is_female</cite>.</p></li>
<li><p><strong>duplicated_samples</strong> – All duplicated samples TO REMOVE (If not provided, this function won’t work as it assumes that each child has exactly two parents)</p></li>
<li><p><strong>i_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Column containing the 1st sample of the pair in the relationship table</p></li>
<li><p><strong>j_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Column containing the 2nd sample of the pair in the relationship table</p></li>
<li><p><strong>relationship_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Column contatining the relationship for the sample pair as defined in this module constants.</p></li>
<li><p><strong>duplicate_samples_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/genetics/hail.genetics.Pedigree.html#hail.genetics.Pedigree" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Pedigree</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Pedigree of complete trios</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.sample_qc.relatedness.create_fake_pedigree">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">create_fake_pedigree</code><span class="sig-paren">(</span><em class="sig-param">n</em>, <em class="sig-param">sample_list</em>, <em class="sig-param">exclude_real_probands=False</em>, <em class="sig-param">max_tries=10</em>, <em class="sig-param">real_pedigree=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/sample_qc/relatedness.html#create_fake_pedigree"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.sample_qc.relatedness.create_fake_pedigree" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a pedigree made of trios created by sampling 3 random samples in the sample list.</p>
<ul class="simple">
<li><p>If <cite>real_pedigree</cite> is given, then children in the resulting fake trios will not include any trio with proband - parents
that are in the real ones.</p></li>
<li><p>Each sample can be used only once as a proband in the resulting trios.</p></li>
<li><p>Sex of probands in fake trios is random.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Number of fake trios desired in the pedigree</p></li>
<li><p><strong>sample_list</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List of samples</p></li>
<li><p><strong>exclude_real_probands</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If set, then fake trios probands cannot be in the real trios probands.</p></li>
<li><p><strong>max_tries</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Maximum number of sampling to try before bailing out (preventing infinite loop if <cite>n</cite> is too large w.r.t. the number of samples)</p></li>
<li><p><strong>real_pedigree</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/genetics/hail.genetics.Pedigree.html#hail.genetics.Pedigree" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Pedigree</span></code></a>]) – Optional pedigree to exclude children from</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/genetics/hail.genetics.Pedigree.html#hail.genetics.Pedigree" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Pedigree</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Fake pedigree</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.sample_qc.relatedness.compute_related_samples_to_drop">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">compute_related_samples_to_drop</code><span class="sig-paren">(</span><em class="sig-param">relatedness_ht</em>, <em class="sig-param">rank_ht</em>, <em class="sig-param">kin_threshold</em>, <em class="sig-param">filtered_samples=None</em>, <em class="sig-param">min_related_hard_filter=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/sample_qc/relatedness.html#compute_related_samples_to_drop"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.sample_qc.relatedness.compute_related_samples_to_drop" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute a Table with the list of samples to drop (and their global rank) to get the maximal independent set of unrelated samples.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p><cite>relatedness_ht</cite> should be keyed by exactly two fields of the same type, identifying the pair of samples for each row.</p></li>
<li><p><cite>rank_ht</cite> should be keyed by a single key of the same type as a single sample identifier in <cite>relatedness_ht</cite>.</p></li>
</ul>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>relatedness_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – relatedness HT, as produced by e.g. pc-relate</p></li>
<li><p><strong>kin_threshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Kinship threshold to consider two samples as related</p></li>
<li><p><strong>rank_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table with a global rank for each sample (smaller is preferred)</p></li>
<li><p><strong>filtered_samples</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.SetExpression.html#hail.expr.SetExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SetExpression</span></code></a>]) – An optional set of samples to exclude (e.g. these samples were hard-filtered)  These samples will then appear in the resulting samples to drop.</p></li>
<li><p><strong>min_related_hard_filter</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – If provided, any sample that is related to more samples than this parameter will be filtered prior to computing the maximal independent set and appear in the results.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A Table with the list of the samples to drop along with their rank.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.sample_qc.relatedness.filter_mt_to_trios">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">filter_mt_to_trios</code><span class="sig-paren">(</span><em class="sig-param">mt</em>, <em class="sig-param">fam_ht</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/sample_qc/relatedness.html#filter_mt_to_trios"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.sample_qc.relatedness.filter_mt_to_trios" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter a MatrixTable to a set of trios in <cite>fam_ht</cite> and annotates with adj.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – A Matrix Table to filter to only trios</p></li>
<li><p><strong>fam_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – A Table of trios to filter to, loaded using <cite>hl.import_fam</cite></p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A MT filtered to trios and adj annotated</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.sample_qc.relatedness.generate_trio_stats_expr">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">generate_trio_stats_expr</code><span class="sig-paren">(</span><em class="sig-param">trio_mt</em>, <em class="sig-param">transmitted_strata={'raw': True}</em>, <em class="sig-param">de_novo_strata={'raw': True}</em>, <em class="sig-param">ac_strata={'raw': True}</em>, <em class="sig-param">proband_is_female_expr=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/sample_qc/relatedness.html#generate_trio_stats_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.sample_qc.relatedness.generate_trio_stats_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a row-wise expression containing trio transmission stats.</p>
<dl class="simple">
<dt>The expression will generate the following counts:</dt><dd><ul class="simple">
<li><p>Number of alleles in het parents transmitted to the proband</p></li>
<li><p>Number of alleles in het parents not transmitted to the proband</p></li>
<li><p>Number of de novo mutations</p></li>
<li><p>Parent allele count</p></li>
<li><p>Proband allele count</p></li>
</ul>
</dd>
</dl>
<p>Transmission and de novo mutation metrics and allele counts can be stratified using additional filters.
<cite>transmitted_strata</cite>, <cite>de_novo_strata</cite>, and <cite>ac_strata</cite> all expect a dictionary of filtering expressions keyed
by their desired suffix to append for labeling. The default will perform counts using all genotypes and append
‘raw’ to the label.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Expects that <cite>mt</cite> is dense if dealing with a sparse MT <cite>hl.experimental.densify</cite> must be run first.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>trio_mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – A trio standard trio MT (with the format as produced by hail.methods.trio_matrix)</p></li>
<li><p><strong>transmitted_strata</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – Strata for the transmission counts</p></li>
<li><p><strong>de_novo_strata</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – Strata for the de novo counts</p></li>
<li><p><strong>ac_strata</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – Strata for the parent and child allele counts</p></li>
<li><p><strong>proband_is_female_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – An optional expression giving the sex the proband. If not given, DNMs are only computed for autosomes.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An expression with the counts</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.sample_qc.relatedness.generate_sib_stats_expr">
<code class="sig-prename descclassname">gnomad.sample_qc.relatedness.</code><code class="sig-name descname">generate_sib_stats_expr</code><span class="sig-paren">(</span><em class="sig-param">mt</em>, <em class="sig-param">sib_ht</em>, <em class="sig-param">i_col='i'</em>, <em class="sig-param">j_col='j'</em>, <em class="sig-param">strata={'raw': True}</em>, <em class="sig-param">is_female=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/sample_qc/relatedness.html#generate_sib_stats_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.sample_qc.relatedness.generate_sib_stats_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a row-wise expression containing the number of alternate alleles in common between sibling pairs.</p>
<p>The sibling sharing counts can be stratified using additional filters using <cite>stata</cite>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function expects that the <cite>mt</cite> has either been split or filtered to only bi-allelics
If a sample has multiple sibling pairs, only one pair will be counted</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input matrix table</p></li>
<li><p><strong>sib_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table defining sibling pairs with one sample in a col (<cite>i_col</cite>) and the second in another col (<cite>j_col</cite>)</p></li>
<li><p><strong>i_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Column containing the 1st sample of the pair in the relationship table</p></li>
<li><p><strong>j_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Column containing the 2nd sample of the pair in the relationship table</p></li>
<li><p><strong>strata</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – Dict with additional strata to use when computing shared sibling variant counts</p></li>
<li><p><strong>is_female</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – An optional column in mt giving the sample sex. If not given, counts are only computed for autosomes.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A Table with the sibling shared variant counts</p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="platform.html" class="btn btn-neutral float-left" title="gnomad.sample_qc.platform" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="sex.html" class="btn btn-neutral float-right" title="gnomad.sample_qc.sex" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>